dev-php/phpunit: version bump, fix bug 639656

kitsunenokenja opened the pull request:
Add ebuild for current stable release from project upstream.

Closes: https://bugs.gentoo.org/639656
Package-Manager: Portage-2.3.13, Repoman-2.3.3

orlitzky commented:
That "phar" file is a zip file containing not only PHPUnit, but also a copy of every library it uses (and every library that those library use...). To package this properly, you'll have to create separate packages (i.e. ebuilds) for them, and then ultimately have PHPUnit-6.x depend on those packages.

@GuillaumeSeren is working on this, so you'll probably want to talk to him if you're interested in that (much larger) task.

kitsunenokenja commented:
@orlitzky, thanks for exposing this fact. It was not directed to my attention when I checked the upstream resources. Their manual only refers to PHP itself and components of it that must be present, neglecting to elaborate that their PHAR is a collection beyond their own sources.

@GuillaumeSeren could you please comment on this PR? It would be counterintuitive for me to repeat everything you've done thus far. The absence of >phpunit-6 in portage is a dependency block for another ebuild I want to introduce, so I'm looking to have this one cleared up.

GuillaumeSeren commented:
@kitsunenokenja Hello,
packaging big php component can take a long time but, it can be done I managed to finish composer,
and I would like to finish full packaged phpunit as well, so I could add tests feature to composer.

If want to help working on that subject it could be great,
the thing is we cant merge half way, so we need to package all deps of phpunit and so on (+ tests when we can).

To track my progress I try to keep my overlay up to date (https://github.com/GuillaumeSeren/gentoo-overlay), but I have been working on others subjects and this is not moving fast so again if want to contribute on this I can review you work and help if you have questions.

To have some more precise idea about the work itself I suggest you read the composer issue in the BTS 439206 (https://bugs.gentoo.org/439206) and more precisely those commits
```git
d2730461386 dev-php/composer: Add 1.2.2 version [Guillaume Seren]                                                                                                                                                                              
bc36d09d052 dev-php/symfony-console: Add 2.7.9 version [Guillaume Seren]                                                                                                                                                                       
2283ea23104 dev-php/symfony-event-dispatcher: Add 2.1.0 version [Guillaume Seren]                                                                                                                                                              
9963181de94 dev-php/symfony-dependency-injection: Add 2.1.0 version [Guillaume Seren]                                                                                                                                                          
e751acece31 dev-php/symfony-yaml: Add 2.1.0 version [Guillaume Seren]                                                                                                                                                                          
d10d7e2f7e3 dev-php/symfony-config: Add 2.1.0 version [Guillaume Seren]                                                                                                                                                                        
7597666e5c4 dev-php/ca-bundle: Add 1.0.0 version [Guillaume Seren]                                                                                                                                                                             
0bccd72e3e2 dev-php/semver: Add 1.0.0 version [Guillaume Seren]                                                                                                                                                                                
a78020686e9 dev-php/spdx-licenses: Add 1.0.0 version [Guillaume Seren]                                                                                                                                                                         
033f5d398db dev-php/jsonlint: Add 1.4.0 version [Guillaume Seren]                                                                                                                                                                              
79a2aeb43a0 dev-php/json-schema: Add 2.0.0 version [Guillaume Seren]                                                                                                                                                                           
d3bd35121b5 dev-php/cli-prompt: Add 1.0.0 version [Guillaume Seren]                                                                                                                                                                            
ea07d004365 dev-php/phar-utils: Add 1.0.0 version [Guillaume Seren]                                                                                                                                                                            
57de338f5b6 dev-php/symfony-finder: Add 2.7.20 version [Guillaume Seren]                                                                                                                                                                       
bdc66737b4f dev-php/symfony-filesystem: Add 2.7.20 version [Guillaume Seren]                                                                                                                                                                   
a7733e14f60 dev-php/symfony-process: Add 2.8.12 version [Guillaume Seren]                                                                                                                                                                      
6c369a57c2e dev-php/psr-log: Add 1.0.2 version [Guillaume Seren]                                                                                                                                                                               
38e69d1f39f dev-php/fedora-autoloader: Add 0.2.1 version [Guillaume Seren]
```

kitsunenokenja commented:
@GuillaumeSeren, excellent. Portage already has a handful of the dependencies that phpunit calls for, and I also see some of them in your overlay. By incorporating what you have thus far I could round this up faster. I'm building out a list of all the deps before determining which ebuilds we're going to need. Some will be version bumps and a small number will be new ebuilds, I think, based on first glance.

Also, unless there is valid objection, I'd like to have a USE flag control the option to just install the PHAR which would bypass all the dependencies. That way users can install normally by default or opt into taking upstream's packaged PHAR.

orlitzky commented:
> Also, unless there is valid objection, I'd like to have a USE flag control the option to just install the PHAR which would bypass all the dependencies. 

This is not quite against policy, but it is strongly discouraged; see e.g. https://bugs.gentoo.org/show_bug.cgi?id=bundled-libs

The main reason is that fixes to the dependencies in the tree would not affect the bundled version of PHPUnit. That's especially bad for security fixes, where we put out announcements like "upgrade to version 100 and you'll be safe!" End users might upgrade the library to the safe v100, not realizing that PHPUnit has its own vulnerable copy of v99 (and there's no way for us to track that).

Another reason is "symbol collisions," where you wind up with a name clash. PHPUnit can be used as a library, so imagine I've written a program that uses PHPUnit to read some test cases. If my program _also_ uses one of the libraries that is bundled into PHPUnit, then we might get a situation where PHPUnit tries to load one version of a class, but another version is already in scope because my program used it. [Bug 471682)](https://bugs.gentoo.org/471682) is an example of that. The phar file might not have this problem if it overloads the include order, I don't know off the top of my head.

That said, there are exceptions to the rule. The PHP runtime itself has things like `media-libs/gd` bundled, but the bundled copy is heavily modified, and it would be incredibly hard to fix at this point. And the PHP project puts out separate security bulletins for their bundled copy, avoiding (well, with twice as much work) that particular problem.

In this case the bundled libraries are 100% vanilla copies, though, so I'd rather not go down that road again.

kitsunenokenja commented:
Six more dependencies to address based on my last count, plus the revision of the phpunit-6.5.5.ebuild itself to go.

kitsunenokenja commented:
> it is strongly discouraged

Very good points. Safe to say I won't include the PHAR as an option in the ebuild. Thank you for the details!

GuillaumeSeren commented:
@kitsunenokenja 
Having a common list could be great and we could use my overlay if you want to host the work in progress and share it easily (or setup one for you).
I am happy to share the amount of work that way we can be faster and make more packages.

About the USE flags to get a phar version, the subject as been discussed before, and to my understanding we try to avoid a phar version, at least in term of control the deps need to build it,
but @orlitzky can probably say more about that than me (and you also have that covered in the composer isssue linked before).

kitsunenokenja commented:
@GuillaumeSeren, can you address the phpunit deps `phar-io/manifest` and `phar-io/version`? They are probably very simple. Also, `theseer/tokenizer` as you already did 2 from theseer before, so this one will also be quick for you I'm sure.

I have all the other `sebastian-*` deps ready and will make use of your overlay to take care of the phpunit directly-related deps next, which would be: `php-code-coverage`, `php-php-token-stream`, `php-file-iterator`, `php-timer`

GuillaumeSeren commented:
@kitsunenokenja 
Yes I think I could, feel free to PR your stuff in my overlay that way we can track the progress.

kitsunenokenja commented:
@GuillaumeSeren you have some packages whose entry names do not match Portage's existing classifications for the same things. It was actually quicker to do the version bumps on everything that Portage already knew about based on what's already in the tree. There will be extra overhead to make a PR of this for your overlay as a result which is counterintuitive, I'm sure you'll understand.

If you want to work through it in the other direction and update your overlay accordingly in order to account for the commits in my branch instead if you feel it would help?

GuillaumeSeren commented:
@kitsunenokenja 
I am working full time right now and I have not enough time to help on packaging so maybe I will just help in the review, as you seems to move forward pretty quick.

orlitzky commented:
> Another reason is "symbol collisions," where you wind up with a name clash. PHPUnit can be used as a library... The phar file might not have this problem if it overloads the include order, I don't know off the top of my head.

To answer myself, this question doesn't even make sense. Although PHPUnit can be used as a library, the phar file can't. So it's not at risk of symbol collisions, but only because there's a bigger problem: other applications have no way to `include 'PHPUnit/whatever.php';` from the phar. We have at least one package doing that, `dev-php/phing`.

kitsunenokenja commented:
Yes if you were to only include the libraries ordinarily but not the PHAR, you shouldn't have a conflict. But you did mention a security concern and it would be very obscure to track PHAR-based ebuilds in that regard to ensure everyone inherits updated versions. That alone is reason enough to not offer the option.

I'll carry on and wrap up the whole PR myself and ping you, orlitzky, when I'm ready. I realised some of the ebuilds reference the PHP deps without considering PHP 7.2 which is valid, so I will also be going back to update those as well.

orlitzky commented:
A lot of these packages have the HOMEPAGE set to http://phpunit.de, but that now redirects to https://phpunit.de/ (with an 's'). Might as well update them to save the environment =)

kitsunenokenja commented:
Good catch, thanks. Took care of the URLs. Had a minor mistake with dependency resolution that portage pointed out, fixed it, now the dep tree builds out right without issue. I'll be testing soon as I expect I may not have set all the autoload.php files right just yet.

Then I presume I'll have to do commit squashing to simplify the history. Almost there. This task hasn't been awfully tremendous. There are a lot of little ebuilds involved sure, but it's straightforward for them all.

kitsunenokenja commented:
@orlitzky, please have a look at the current state of the PR. The last set of changes are set, and I've installed and successfully run phpunit. Looking to make any outstanding corrections first before squashing everything.

orlitzky commented:
I don't see anything wrong, can you please squash everything?

Once the commits look nice, I'll merge then one at a time and test them fully. Then if I get halfway through the list of commits and find a problem, we don't have to start over -- you can just rebase onto master.

kitsunenokenja commented:
@orlitzky, rebase complete.

orlitzky commented:
> Unless the decision is to drop the old 5.7 ebuild then this will have to be fixed by moving to a separate file to resolve the conflict.

The 5.x ebuild is stable, so we can't drop it when 6.x is added.

orlitzky commented:
The autoloader for theseer-tokenizer isn't working =(

You can fix it by changing the entries to something like,

```php
'theseer\tokenizer\exception' => '/Exception.php',
```

Once that's fixed, please install the `README.md` file, and see if you can't get the tests working. I was able to get them to run by copying `autoload.php` into the `src` directory and then running "phpunit". Note that if you put `autoload.php` into `src`, then you won't have to `doins` it later on because you already install `src/*`.

orlitzky commented:
Oh and I did already pull the first commit, so you should be able to rebase it away.

orlitzky commented:
Just pulled 4b9ff85 (File_Iterator) too.

kitsunenokenja commented:
Confirmed that `dev-php/myclabs-deepcopy` and `dev-php/File_Iterator` were merged into the mainline, and removed them via rebase from my working branch.

Committed the fix and rebased regarding the autoload file for `dev-php/theseer-tokenizer`.

@orlitzky, regarding installing README.md and making test entries in the ebuild, is that a generalised request for all the remaining packages or just for `dev-php/theseer-tokenizer`? Lastly I'm not sure about the suggestion for handling autoload.php in the ebuild. I'd need more instructions surely, to drop it into src/ first since it will have to merge into the vendor subdirectory of the final destination path which ordinarily doesn't exist. The current instruction for installing this file where it belongs is probably the most intuitive and simplest. I don't mind making the change; it just doesn't seem like an improvement. Correct me if I'm wrong.

orlitzky commented:
We should have the docs (like README.md) for everything, because that's pretty easy to do and the benefit is high. _Some_ test suites are easy to get working, but others are a pain in the ass -- it depends on which type of autoloader it uses IIRC. I would prefer to have the easy ones enabled, but if it requires a whole separate autoloader and phpunit.xml patching and things like that, then don't bother.

For `autoload.php` in theseer-tokenizer, I just meant that in order to run the test suite, you need to use the autoloader. If you look in its `phpunit.xml`, you'll see...

```xml
bootstrap="src/autoload.php"
```

and without that file, running "phpunit" will crash. So somewhere in e.g. `src_prepare`, you would need  to copy `${FILESDIR}/autoload.php` into the `src` directory, where phpunit can find it. But then later, there's no reason to `doins ${FILESDIR}/autoload.php`, because `doins src/*` will already find it.

If I'm still talking gibberish, I can try to add the test suite for theseer-tokenizer to show you what I mean.

orlitzky commented:
Any reason not to use v2.0.2 of PHP_TokenStream?

kitsunenokenja commented:
I follow you; even though I just pushed a rebased commit for that I'll do it again to address the autoload.php file. At least for now you can double-check my voluntary usage of USE="doc" before I start doing the same pattern for all the others that have a README to be installed (many if not all probably do, as you said earlier.)

Regarding versioning, I went by the required deps as listed in the composer/packagist site so that everything is satisfactory. I can do an ebuild for 2.0.2 for PHP_TokenStream too. A small number of these have alternate versions already available that are higher than what is strictly required for phpunit 6.5.5.

orlitzky commented:
> double-check my voluntary usage of USE="doc" before I start doing the same pattern for all the others

This is not at all obvious, but the "doc" flag isn't meant to be used like that. Regular end-user documentation like README files, changelogs, man pages, and so on are all supposed to be installed unconditionally. The "doc" flag is used only when there's some other big documentation component, like API documentation or PDFs that need to be built.

I'm certain this policy is written down _somewhere_, but I can't find it right now, so you'll have to believe the description of the flag from use.desc:

> doc - Add extra documentation (API, Javadoc, etc). It is recommended to enable per package instead of globally

kitsunenokenja commented:
That's what I asked, thanks. I had some doubt about whether the description of USE="doc" would apply here or not.

My src_test() is doing what it should but I'm seeing the test fail. I think I know why. Going to experiment a bit first.

kitsunenokenja commented:
There are previous versions of `PHP_TokenStream` and they never supported the test suite. phpunit is launching correctly but the first test dies because the autoloader didn't supply information on where to find all the token classes (the one phpunit dies with is `PHP_Token_OPEN_TAG` class.)

I suppose if I load up this huge list of `PHP_Token` class derviatives the test suite will work. Now I know what you meant by "others are a pain in the ass."

kitsunenokenja commented:
`dodoc` lines added and rebased

orlitzky commented:
Don't worry too much about the test suites that give you trouble, especially not the ones for existing packages. If there's an existing package without a test suite, then it's probably because somebody realized it was going to be a PiTA and skipped it.

Here's an example for theseer-tokenizer, though, to show you what I mean. This runs and passes with FEATURES=test now (using phpunit-5.x, naturally):

```sh
# Copyright 1999-2018 Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2

EAPI=6

DESCRIPTION="Convert tokenized PHP source code into XML and other formats"
HOMEPAGE="https://github.com/theseer/tokenizer"
SRC_URI="${HOMEPAGE}/archive/${PV}.tar.gz -> ${P}.tar.gz"

LICENSE="BSD"
SLOT="0"
KEYWORDS="~alpha ~amd64 ~arm ~hppa ~ia64 ~ppc ~ppc64 ~s390 ~sh ~sparc ~x86"
IUSE="test"

S="${WORKDIR}/tokenizer-${PV}"

RDEPEND="dev-php/fedora-autoloader
	>=dev-lang/php-7.0:*[tokenizer,xmlwriter]"

DEPEND="test? (	${RDEPEND} dev-php/phpunit )"

src_prepare() {
	# This ultimately needs to live next to the other src/*.php files,
	# but it needs to be there *now* if we're going to run the tests.
	cp "${FILESDIR}/autoload.php" src/ || die
	default
}

src_install() {
	insinto /usr/share/php/TheSeer/Tokenizer
	doins src/*.php
	dodoc README.md
}

src_test() {
	phpunit || die 'test suite failed'
}
```

kitsunenokenja commented:
Despite still learning how to write ebuilds, the changes I had tested looked very similar to that. I didn't use `src_prepare` which is better but the effect was the same. That aside, the other necessary changes are pending; if anything is still not ready to be merged, just show the way and I'll fix it up.

orlitzky commented:
Do you want to update your copy of theseer-tokenizer? I'd rather give you credit for the commit than use my own.

kitsunenokenja commented:
Update with what? If we aren't adding the test suite to it, then isn't it already prepared?

orlitzky commented:
The test suite is missing from https://github.com/gentoo/gentoo/pull/6783/commits/53b7899842a78d84a875dbf88cae9d4a47a1c097, am I looking in the right place?

kitsunenokenja commented:
@orlitzky, maybe you can see what I'm not seeing to solve this. I updated the autoload script for `PHP_TokenStream` so all the extra class definitions that are summoned during unit testing will be loaded properly. The unit test phase mostly runs but fails to load PHPUnit's MockBuilder class. I've traced the flow and found that mock's autoload.php file does get pulled in and runs, which defines how to locate `PHPUnit\Framework\MockObject\MockBuilder` yet the unit testing still fails with an error saying it didn't find that class. I'm not sure why that error is happening because, like I tested, the autoloader script explaining where it is, did run, and it looks correct.

orlitzky commented:
I think you should skip the tests for PHP_TokenStream for now (but do use v2.0.2 instead of 2.0.1). Its test suite requires phpunit-6.x, and it's kinda hard to test that until after we get this PR merged.

kitsunenokenja commented:
The change was a necessity to write the test for `TheSeer/Tokenizer`. I wasn't writing a test routine for the `PHP_TokenStream` package itself. Shall I add Token Stream's 2.0.2 ebuild into the PR?

orlitzky commented:
Oh, your https://github.com/gentoo/gentoo/pull/6783#issuecomment-359963974 made me think it was PHP_TokenStream that was a problem. The ebuild in https://github.com/gentoo/gentoo/pull/6783#issuecomment-357850947 works for me with your existing `autoload.php` and the other ebuilds (already merged) that are in the tree.

Are you running the test suite with phpunit-6.x? I was using 5.x, maybe that's the difference?

orlitzky commented:
And yeah, I would go ahead and use tokenstream-2.0.2, it just contains one small bugfix over v2.0.1.

kitsunenokenja commented:
Replaced 2.0.1 with 2.0.2. And yes I am using phpunit 6.5.5; every package in this PR is installed on my box.

orlitzky commented:
> And yes I am using phpunit 6.5.5

Ah, ok, then in my theseer-tokenizer example you'll want to change

```sh
DEPEND="test? (	${RDEPEND} dev-php/phpunit )"
```

to

```sh
DEPEND="test? (	${RDEPEND} dev-php/phpunit-5* )"
```

to ensure that its tests get run with phpunit-5.x

kitsunenokenja commented:
That sounds really counterintuitive. A user that wants to run the test suite will have to take a lower version to do so and then ultimately replace it with 6.x afterwards if the original intent is to both have 6.x and run the tests? Are we clear this is a problem with the test for that package not cooperating with 6.x rather than there being a mistake causing the issue? I will make this change but it seems worth asking if it's the right path to take.

orlitzky commented:
Yeah, the test suite simply fails with phpunit-6.x (I tried with the phar releases for both versions). It's crap, but the situation with the other ecosystems is pretty much the same. Anything that's a dependency of the test runner program itself will have a USE=test that causes circular dependencies and conflicts and all other manner of horror. For the same reason, building the API docs (with USE=doc) for the language's documentation generator usually causes chaos.

End users would be wise to leave FEATURES=test disabled for such things, but it will come in real handy in the future when we want to do an update to *only* theseer-tokenizer (or packages like it) and want to make sure that it will function with (say) php-7.3.

orlitzky commented:
> Yeah, the test suite simply fails with phpunit-6.x

Scratch that, I was thinking about PHP_TokenStream, not theseer-tokenizer, and I had it backwards. Too many damn "token"s =)

The PHP_TokenStream suite only works with PHPUnit-6.x. That will cause circular dependencies, but you can still make it work by installing PHPUnit-6.x with FEATURES="-test" once and then enabling FEATURES=test.

The theseer-tokenizer test suite looks like it **does** work with phpunit-6.x, which means... the autoload.php for some intermediate package is probably not working right. Go back to phpunit-5.x (from ::gentoo) if you can, because we know that works at least. Then once I get all this stuff merged we can check the test suites with phpunit-6.x.

kitsunenokenja commented:
Change is in, and the cyclic issues are another reason to refrain from supporting testing. If it can be cleaned up I'm all right with further tuning the ebuilds accordingly.

orlitzky commented on dev-php/theseer-tokenizer/theseer-tokenizer-1.1.0.ebuild:
```diff
@@ -0,0 +1,37 @@
+RDEPEND="dev-php/fedora-autoloader
+	>=dev-lang/php-7.0:*[tokenizer,xmlwriter]"
+
+DEPEND="test? ( ${RDEPEND} =dev-php/phpunit-5* )"
```
I was wrong about this, there's no need for the `-5*`

orlitzky commented on dev-php/theseer-tokenizer/theseer-tokenizer-1.1.0.ebuild:
```diff
@@ -0,0 +1,37 @@
+DEPEND="test? ( ${RDEPEND} =dev-php/phpunit-5* )"
+
+src_prepare() {
+	cp "${FILESDIR}/autoload.php" src/
```
This line needs `|| die` at the end

orlitzky commented on dev-php/theseer-tokenizer/files/autoload.php:
```diff
@@ -0,0 +1,20 @@
+/* Autoloader for dev-php/theseer-tokenizer */
+
+if (!class_exists('Fedora\\Autoloader\\Autoload', false)) {
+	require_once '/usr/share/php/Fedora/Autoloader/autoload.php';
```
I think these would actually be better without the `/usr/share/php`, so instead...

```php
        require_once 'Fedora/Autoloader/autoload.php';
```

(thats lets us play games with the `include_dir` and override the default stuff with version-specific copies).

kitsunenokenja commented:
Good suggestions. Thanks. They're in now.

orlitzky commented:
I merged theseer-tokenizer so you can rebase that one away. Now it looks like the autoload.php for phar-io-version doesn't work when trying to run the example from https://github.com/phar-io/version/blob/master/README.md

NB: we should probably install into `PharIo/Version` rather than `phar-io/version`, to match the layout of the other PSR autoloaders.

kitsunenokenja commented:
Having seen these patterns more often, now I've come to know what to expect. I took care of both `phar-io-version` and `phar-io-manifest` regarding both the autoloader entries and the install path in the ebuild.

orlitzky commented:
Still no luck with the phar-io-version demo...

```
$ php phar-io-version-test.php 
PHP Fatal error:  Uncaught Error: Class 'PharIo\Version\VersionConstraintParser' not found in /home/mjo/tmp/phpunit-bump/phar-io-version-test.php:8
Stack trace:
#0 {main}
  thrown in /home/mjo/tmp/phpunit-bump/phar-io-version-test.php on line 8
```

The contents of that file are...

```php
<?php

require 'PharIo/Version/autoload.php';

use PharIo\Version\Version;
use PharIo\Version\VersionConstraintParser;

$parser = new VersionConstraintParser();
$caret_constraint = $parser->parse( '^7.0' );

echo $caret_constraint->complies( new Version( '7.0.17' ) ); // true
echo $caret_constraint->complies( new Version( '7.1.0' ) ); // true
echo $caret_constraint->complies( new Version( '6.4.34' ) ); // false

echo $tilde_constraint = $parser->parse( '~1.1.0' );

echo $tilde_constraint->complies( new Version( '1.1.4' ) ); // true
echo $tilde_constraint->complies( new Version( '1.2.0' ) ); // false
```

kitsunenokenja commented:
@orlitzky, seems those underscores don't work in the autoloader script and I'll just stick to the proper backslash delimiters. I had seen it somewhere else and figured those strings would be okay that way. Well, now we know.

Also, your test code has a bug. You can't call echo with the result of the parse method because it returns an object. Please use this revised version; it is fixed and passed the test:

```php
<?php
require_once '/usr/share/php/PharIo/Version/autoload.php';

use PharIo\Version\Version;
use PharIo\Version\VersionConstraintParser;

$parser = new VersionConstraintParser();
$caret_constraint = $parser->parse( '^7.0' );

var_dump($caret_constraint->complies( new Version( '7.0.17' ) )); // true
var_dump($caret_constraint->complies( new Version( '7.1.0' ) )); // true
var_dump($caret_constraint->complies( new Version( '6.4.34' ) )); // false

$tilde_constraint = $parser->parse( '~1.1.0' );

var_dump($tilde_constraint->complies( new Version( '1.1.4' ) )); // true
var_dump($tilde_constraint->complies( new Version( '1.2.0' ) )); // false
```

orlitzky commented:
In hindsight, the PharIo/Version autoloader can be a lot simpler. The "classMap" autoloader is only needed for the packages that have a nonstandard layout. The PharIo/Version scheme maps classes to files according to PSR-4, so the following should suffice:

```php
<?php
/* Autoloader for dev-php/phar-io-version */

require_once 'Fedora/Autoloader/autoload.php';

\Fedora\Autoloader\Autoload::addPsr4('PharIo\\Version\\', __DIR__);
```

Some of the others are probably similar. Sorry for not noticing that earlier!

orlitzky commented:
Ok, I merged phar-io-version, but now the autoloader for phar-io-manifest needs work =)

It's at least missing the dependency on the autoloader for phar-io-version that will look something like this (untested):

```php
\Fedora\Autoloader\Dependencies::required(array(
    '/usr/share/php/PharIo/Version/autoload.php',
));
```

Afterwards please check that the example in the README at https://github.com/phar-io/manifest can be run.

orlitzky commented:
Cool, phar-io-manifest is merged. Some of the remaining commits need to be reordered; the new PHP_CodeCoverage for example depends on the new sebastian-environment-3.0.0, which I think is next on my list. Which brings me to... the fact that we need a new `autoload.php` for the new sebastian-environment. There's a new file in `OperatingSystem.php` that isn't addressed in the current `autoload.php`. Since the old sebastian-environment-2.0.0 is marked stable on a few arches, we have to make a new one and not modify the old one.

Some good news, though: I think we can use the `addPsr4()` method in the new sebastian-environment autoloader, so that this problem won't happen again.

kitsunenokenja commented:
`phar-io-manifest` is removed from PR, and `PHP_CodeCoverage` is pushed down to come after the group of sebastian-* packages. I've put in the addPsr4 call in the new autoloader in its own file, autoloader-3.1.0.php, the same way I addressed the same issue with phpunit's ebuild.

I wrote this as a quick way to verify the autoloader is resolving properly.

```php
<?php
require_once "/usr/share/php/SebastianBergmann/Environment/autoload.php";
use SebastianBergmann\Environment\OperatingSystem;

echo (new OperatingSystem())->getFamily() . "\n";
```

orlitzky commented:
> I've put in the addPsr4 call in the new autoloader in its own file, autoloader-3.1.0.php, the same way I addressed the same issue with phpunit's ebuild.

Thanks, it looks like we'll need to do the same for the PHP_TokenStream autoloader because the existing one is used by the stable PHP_TokenStream-1.4.11.

kitsunenokenja commented:
Both 1.4 and 2.0 branches for PHP_TokenStream appear to have classes defined in the same way in the same files. It did not appear that a different file should be made or is needed for the 2.0 ebuild. Did you notice any differences?

orlitzky commented:
There may not be any differences (in other words, the test class additions simply fix a bug in the old autoloader), but since the 1.4 version is stable we're not supposed to touch the ebuild or any of its files. "Stable" in Gentoo means "it doesn't change," not "it works right" =)

kitsunenokenja commented:
I've been staring at the current version of that particular autoloader so long now, that I've forgotten it isn't the original version of that file. Sorry about that. I'll clean up that part of the PR then and restore the original autoload.php that's associated with the stable ebuild already in place.

kitsunenokenja commented:
@orlitzky, could you provide a status update please?

orlitzky commented:
@kitsunenokenja I've been working double shifts at my dayjob and have lost the will to live, a prerequisite for Gentoo work. It will be at least another week before this project is done =(

kitsunenokenja commented:
@orlitzky, all right, take your time then. Just checking that this didn't fall through the cracks since this has been a larger undertaking than most PRs, didn't want it to become forgotten! Thanks.

orlitzky commented:
My work situation hasn't changed... you'd better ask around, sorry to drop the ball.
